cmd/spire-server/cli/agent/list.go

```diff
@@ -40,11 +41,19 @@ func (listCommand) Synopsis() string {
 // Run lists attested agents
 func (c *listCommand) Run(ctx context.Context, env *common_cli.Env, serverClient util.ServerClient) error {
 	agentClient := serverClient.NewAgentClient()
-	listResponse, err := agentClient.ListAgents(ctx, &agentv1.ListAgentsRequest{})
+	listResponse, err := agentClient.ListAgents(ctx, &agentv1.ListAgentsRequest{
+		PageSize: int32(unsafe.Sizeof(*(*types.Agent)(nil))),
```

Comment:
Not sure about the `Sizeof` business - according to proto description, this should be number of entries on the page?

Reply:
https://github.com/spiffe/spire/issues/2443#issuecomment-897259993
Did I get this wrong?

Reply:
Not sure what was your intent. I modified in https://github.com/transferwise/spire/commit/178f377d5dc8528e24dc543b6974e55e42f4a0dc#diff-3293f78c44787793c3ffc2b7ab709de4f2c0d73e53f2fe43a7d43361706ede00 and that works (also added looping while there are pages left).

Why would you choose as PageSize the in-memory size of some structure? Those are unrelated. Structure size is what it takes in memory in bytes, PageSize is how many entries you would like per response.

(Note: I'm not a spire contributor, mostly have no idea what I'm talking about, take comments with grain of salt).

Reply:
Note: consult https://github.com/spiffe/spire-api-sdk/blob/main/proto/spire/api/server/agent/v1/agent.proto#L90 on request/response field interpretation.

Reply:
Yes, I got that I should use a numerical value for page size. But as mentioned in the comment, it should depend on the size of the agent (in this case) and not a hardcoded value. How can use this size of the agent and use it to define the page-size?

Reply:
I would actually prefer we use a hardcoded page size limit here. The value of that limit can be derived from offline calculations about theoretical upper bounds of the size of the `Agent` protobuf when all the fields are populated. What I was trying to suggest in the related issue (and didn't do a very good job of explaining) was that we do some math to estimate how large each field in the response message could be (e.g. think about maximum allowable sizes of each corresponding field in the datastore).

Based on those estimates, it should be pretty straightforward to write a small unit test/application that does roughly the following:

1. Construct a [types.Agent](https://github.com/spiffe/spire-api-sdk/blob/main/proto/spire/api/types/agent.proto#L8) with each field populated with max size values, e.g.
```go
a := types.Agent{
    Id: &types.SPIFFEID{
        TrustDomain: "somereallylongtrustdomainbasedondatastorelimits",
        Path: "somereallylongpathbasedondatastorelimits",
    },
    // Populate remaining fields similarly
    ...
}
```
2. Print max size of that `types.Agent` using the `[]byte` obtained from `proto.Marshal()`
3. Create a [ListAgentsResponse](https://github.com/spiffe/spire-api-sdk/blob/7ff3eb0759ce242a38fd02aa368b3d7a33ac7316/proto/spire/api/server/agent/v1/agent.proto#L96) with the largest conceivable `types.Agent` object created in 1) and a `next_page_token` populated (should be size of UUID).
4. Repeat 2) with the `ListAgentsResponse` created in 3)

From that data, we can derive the approximate max size of a proto-serialized individual Agent, as well as the approximate max size of the API response type `ListAgentsResponse`. Once we know that data, we can derive a safe page size using a conservative percentage of the theoretical maximum number of agents that could fit into a 4 MB response, e.g. 80%.

Does that make sense @sachinkumarsingh092 ? I'm happy to chat offline too if that's easier.

Reply:
@robinp-tw has the right idea. What we're trying to avoid is surpassing the 4MiB default message limit imposed by the gRPC server implementation. What we would like to determine is a rough estimate of the number of agent messages we can fit in a single response without passing that limit. To determine the rough estimate of the maximum size a single Agent message we need to look at each field in the Agent message and guesstimate about how large that field could be when encoded on the wire. The size of the `types.Agent` struct in memory isn't a good approximation, since that gives you number of of bytes the struct that represents the message takes in memory, and considering many of the fields are "pointery" (i.e. the agent ID), it isn't even close to accurate. To get a good estimate, we need to go field by field. For example, SPIFFE ID's are at most 2048 bytes, so we need at least 2K for such a field, etc.

Once you've come up with a rough theoretical maximum, we can divide 4MiB by that number to get a reasonable page size.

For example, if we estimate the agent to be at most 4K per wire (I'm making this number up), then we would land on a page size of 4MiB / 4K == ~1024 pages. Now, to be safe, it would probably be prudent to reduce this page size by some percentage just to make sure we won't exceed the limit. We don't have to be exact here, we just have to be reasonable close to the limit without a chance of going over. Something like 80% would be pretty safe, IMO without being terribly "wasteful".

Reply:
Whoops! I should have refreshed my tab before commenting :) Sorry for the repeat advice :) Thanks @rturner3 !

Reply:
Random data point from a real-world scenario: it took a bit less than 8000 agents to hit the 4MiB limit.

Reply:
How do we get that maximum value for a given field from the datastore? Since we are using gorm, we can have a different database that might have a different maximum size for the unbounded `string` type.

Reply:
We're going to end up being a little imprecise here. SPIRE doesn't currently enforce a maximum, but the SPIFFE specification [says](https://github.com/spiffe/spiffe/blob/main/standards/SPIFFE-ID.md#23-maximum-spiffe-id-length) the following:

> the recommended maximum length of a SPIFFE ID is 2048 bytes.

Assuming a SPIFFE ID is at most 2048 bytes seems fairly safe, practically speaking. A "safe" maximum for the other string fields can also be assumed (e.g. our serial numbers are random 128-bit integers). We can also take a guesstimate of an appropriate upper bound of selectors an agent is likely to have associated with it during node attestation. We don't need to be perfect in our estimations, and should be pretty liberal when guessing at upper bounds. In addition, taking only 80% of the calculated page size like Ryan suggested will help accommodate for some of that uncertainty.

Reply:
I did something like this to find out the maximum of a sample agent:
```
	sample_agent := types.Agent{
		Id: &types.SPIFFEID{
			TrustDomain: "somereallylongtrustdomainbasedondatastorelimits",
			Path:        "somereallylongpathbasedondatastorelimits",
		},

		AttestationType:      "workloadornode",
		X509SvidSerialNumber: "253171134110531215125233123392001534969203",
		X509SvidExpiresAt:    1629304512,
		Selectors: []*types.Selector{
			{Type: "k8s_psat", Value: "agent_ns:spire"},
			{Type: "k8s_psat", Value: "agent_sa:spire-agent"},
			{Type: "k8s_psat", Value: "cluster:demo0-cluster"},
			{Type: "k8s_psat", Value: "cluster:demo1-cluster"},
			{Type: "k8s_psat", Value: "cluster:demo2-cluster"},
			{Type: "k8s_psat", Value: "cluster:demo3-cluster"},
			{Type: "k8s_psat", Value: "cluster:demo4-cluster"},
			{Type: "k8s_psat", Value: "cluster:demo5-cluster"},
		},
		Banned: true,
	}

	enc, err := proto.Marshal(&sample_agent)
	if err != nil {
		return err
	}

	env.Printf("The maximum agent size is %v\n", len(enc))
```
But then, I could still add more characters to make all these even longer. So I started to calculate this theoretically and this is how I allocated the fields:
```
types.Agent{
		Id: 2048 bytes
		AttestationType: 8 bytes,
		X509SvidSerialNumber: 16 bytes,
		X509SvidExpiresAt:  10 bytes,
		Selectors: 32 (size of 1 selector) * 10 (max selectors) bytes
		Banned: 1 byte,
}

NextPage: 16 bytes
```
Is this theoretical maximum acceptable? What can be possible changes?

Reply:
That's probably a decent place to start. Reducing our page limit to 80% of the value we calculate will cover small discrepancies.